File: leetcode/0313-super-ugly-number/Solution.java
class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        int k = primes.length;
        long[] ugly = new long[n];
        ugly[0] = 1;
        int[] pointers = new int[k];

        for (int i = 1; i < n; i++) {
            long nextUgly = Long.MAX_VALUE;
            for (int j = 0; j < k; j++) {
                nextUgly = Math.min(nextUgly, ugly[pointers[j]] * primes[j]);
            }
            ugly[i] = nextUgly;

            for (int j = 0; j < k; j++) {
                if (ugly[pointers[j]] * primes[j] == nextUgly) {
                    pointers[j]++;
                }
            }
        }

        return (int) ugly[n - 1];
    }
}

File: leetcode/0313-super-ugly-number/solution.py
from typing import List


class Solution:
    def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
        k = len(primes)
        ugly = [1]
        pointers = [0] * k

        while len(ugly) < n:
            candidates = [ugly[pointers[i]] * primes[i] for i in range(k)]
            next_ugly = min(candidates)
            ugly.append(next_ugly)

            for i in range(k):
                if candidates[i] == next_ugly:
                    pointers[i] += 1

        return ugly[-1]
